File: mobile/lib/models/customer_category.dart
class CustomerCategory {
  final int id;
  final String name;
  final int? parentId;
  final List<CustomerCategory> children;

  CustomerCategory({
    required this.id,
    required this.name,
    this.parentId,
    this.children = const [],
  });

  bool get isSubcategory => parentId != null;

  factory CustomerCategory.fromJson(Map<String, dynamic> json) {
    return CustomerCategory(
      id: json['id'],
      name: json['name'],
      parentId: json['parent_id'],
      children: json['children'] != null
          ? (json['children'] as List).map((e) => CustomerCategory.fromJson(e)).toList()
          : const [],
    );
  }
}

File: backend/app/Models/CustomerCategory.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class CustomerCategory extends Model
{
    protected $fillable = ['parent_id', 'name', 'description', 'is_active'];

    protected function casts(): array
    {
        return ['is_active' => 'boolean'];
    }

    public function customers(): HasMany
    {
        return $this->hasMany(Customer::class);
    }

    public function parent(): BelongsTo
    {
        return $this->belongsTo(CustomerCategory::class, 'parent_id');
    }

    public function children(): HasMany
    {
        return $this->hasMany(CustomerCategory::class, 'parent_id');
    }

    public function isSubcategory(): bool
    {
        return $this->parent_id !== null;
    }
}
